kamu-core/src/infra/repos/dataset_repository_s3.rs

```diff
@@ -0,0 +1,473 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::{str::FromStr, sync::Arc};
+
+use async_trait::async_trait;
+use dill::*;
+use futures::TryStreamExt;
+use opendatafabric::{DatasetHandle, DatasetName, DatasetRefLocal};
+use rusoto_s3::*;
+use thiserror::Error;
+use url::Url;
+
+use crate::{domain::*, infra::utils::s3_context::S3Context};
+
+use super::{get_downstream_dependencies_impl, get_staging_name, DatasetFactoryImpl};
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[component(pub)]
+pub struct DatasetRepositoryS3 {
+    s3_context: S3Context,
+    endpoint: String,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+impl DatasetRepositoryS3 {
+    pub fn new(s3_context: S3Context, endpoint: String) -> Self {
+        Self {
+            s3_context,
+            endpoint,
+        }
+    }
+
+    fn get_s3_bucket_path(&self, dataset_name: &str) -> Url {
+        let dataset_url_string = format!(
+            "s3+{}/{}/{}/",
+            self.endpoint,
+            self.s3_context.bucket,
+            self.s3_context.get_key(dataset_name)
+        );
+        Url::parse(dataset_url_string.as_str()).unwrap()
+    }
+
+    fn get_dataset_impl(&self, dataset_name: &DatasetName) -> Result<impl Dataset, InternalError> {
+        let dataset_url = self.get_s3_bucket_path(dataset_name.as_str());
+        DatasetFactoryImpl::get_s3(dataset_url)
+    }
+
+    async fn eliminate_bucket_items(
+        &self,
+        dataset_name: &DatasetName,
+    ) -> Result<(), InternalError> {
+        let dataset_key_prefix = self.s3_context.get_key(dataset_name.as_str());
+        let list_response = self
+            .s3_context
+            .client
+            .list_objects(ListObjectsRequest {
+                bucket: self.s3_context.bucket.clone(),
+                prefix: Some(dataset_key_prefix),
+                ..ListObjectsRequest::default()
+            })
+            .await
+            .int_err()?;
+
+        if let Some(contents) = list_response.contents {
+            self.s3_context
+                .client
+                .delete_objects(DeleteObjectsRequest {
```

Comment:
`list_objects` only returns up to 1000 items
likewise `delete_objects` can accept up to 1000 keys
If you don't implement "pagination" then at lease an assertion is in order to detect when we exceeded the limit.

Reply:
Addressed